llama.cpp

```diff
@@ -8398,6 +8489,194 @@ struct llm_build_context {
+            cb(selected_experts->src[0], "ffn_moe_argsort", il);
+
+            ggml_tensor * weights = ggml_get_rows(ctx0,
+                                                  ggml_reshape_3d(ctx0, probs, 1, n_expert, n_tokens), selected_experts);
+            cb(weights, "ffn_moe_weights", il);
+
+            weights = ggml_reshape_2d(ctx0, weights, n_expert_used, n_tokens);  // [n_tokens, num_experts_per_tok]
+
+            // compute expert outputs
+            ggml_tensor * moe_out = nullptr;
+
+            for (int i = 0; i < n_expert_used; ++i) {
+                ggml_tensor * cur_expert;
+
+                ggml_tensor * cur_up = ggml_mul_mat_id(ctx0, model.layers[il].ffn_up_exps, selected_experts, i, cur);
+                cb(cur_up, "ffn_moe_up", il);
+
+                ggml_tensor * cur_gate = ggml_mul_mat_id(ctx0, model.layers[il].ffn_gate_exps, selected_experts, i, cur);
+                cb(cur_gate, "ffn_moe_gate", il);
+
+                cur_gate = ggml_silu(ctx0, cur_gate);
+                cb(cur_gate, "ffn_moe_silu", il);
+
+                cur_expert = ggml_mul(ctx0, cur_up, cur_gate); // [n_tokens, n_embd]
+                cb(cur_expert, "ffn_moe_gate_par", il);
+
+                cur_expert = ggml_mul_mat_id(ctx0, model.layers[il].ffn_down_exps, selected_experts, i, cur_expert); // [n_tokens, n_embd]
+                cb(cur_expert, "ffn_moe_down", il);
+
+                cur_expert = ggml_mul(ctx0, cur_expert,
+                                      ggml_view_2d(ctx0, weights, 1, n_tokens, weights->nb[1], i*weights->nb[0]));
+                cb(cur_expert, "ffn_moe_weighted", il);
+
+                if (i == 0) {
+                    moe_out = cur_expert;
+                } else {
+                    moe_out = ggml_add(ctx0, moe_out, cur_expert);
+                    cb(moe_out, "ffn_moe_out", il);
+                }
+            }
+
+            ggml_tensor * gate_shared_exp = ggml_mul_mat(ctx0, model.layers[il].ffn_gate_inp_shared_exp, cur);
+            cb(gate_shared_exp, "ffn_moe_gate_inp_shared_exp", il);
+
+            // sigmoid
+            ggml_tensor * logits_shared_exp = ggml_silu(ctx0, gate_shared_exp);
+            cb(logits_shared_exp, "ffn_moe_logits_shared_exp", il);
+
+            ggml_tensor * probs_shared_exp = ggml_div(ctx0, logits_shared_exp, gate_shared_exp);
+            cb(probs_shared_exp, "ffn_moe_probs_shared_exp", il);
+
+            ggml_tensor * ffn_shared_exp = llm_build_ffn(ctx0, cur,
+                    model.layers[il].ffn_up_shared_exp,   NULL,
+                    model.layers[il].ffn_gate_shared_exp, NULL,
+                    model.layers[il].ffn_down_shared_exp, NULL,
+                    NULL,
+                    LLM_FFN_SILU, LLM_FFN_PAR, cb, il);
+            cb(ffn_shared_exp, "ffn_moe_shared_exp", il);
+
+            ggml_tensor * ffn_shared_exp_out = ggml_mul(ctx0, ffn_shared_exp, probs_shared_exp);
+            cb(ffn_shared_exp_out, "ffn_moe_shared_exp_out", il);
```

Comment:
Is this meant to be equivalent to this?
https://github.com/bozheng-hit/transformers/blob/d9dc993fdd6e6b0a61fe68ccbe838e00c73b9f80/src/transformers/models/qwen2_moe/modeling_qwen2_moe.py#L860-L863

Reply:
"logits" ∈(−∞,∞) and "probs" ∈[0,1] look good to me. I can not find better options here.